tools/mason/MasonPublish.chpl

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+use MasonUtils;
+use Spawn;
+use FileSystem;
+use TOML;
+use MasonEnv;
+use MasonNew;
+use MasonModify;
+use Random;
+
+/* Top Level procedure that gets called from mason.chpl that takes in arguments from command line
+Returns the help output in '-h' or '--help' exits in the arguments.
+If --dry-run is passed then it checks to see if the package is able to be published.
+Takes in the username of the package owner as an argument*/
+proc masonPublish(args: [] string) throws {
+  try! {
+    if args.size == 4 {
+      masonPublishHelp();
+      exit(0);
+    }
+    var dry = false;
+    var username = getUsername();
+    if username == '' {
```

Comment:
This error message seems like it could be misleading if the url is munched in some way versus it not actually being present. esp given that you are checking for the username here. might be a nit.

Reply:
Git will return a better error depending on what goes wrong with the url. the only case they don't seem to output is when it doesn't exist at all and then it will return an empty string, but a bad username, git takes care of the error message for us